File: 0198-house-robber/0198-house-robber.java
class Solution {
    int[] memo;
    public int rob(int[] nums) {
        memo = new int[nums.length];
        Arrays.fill(memo, -1);
        return Math.max(helper(nums, 0), helper(nums, 1));
    }

    private int helper(int[] nums, int i) {
        if (i > nums.length - 1) return 0;
        if (memo[i] != -1) return memo[i];

        memo[i] = nums[i] + Math.max(helper(nums, i + 2), helper(nums, i + 3));
        return memo[i];
    }
}

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prev1{}, prev2{};

        for (int num : nums) {
            auto tmp = prev1;
            prev1 = std::max(prev1, prev2 + num);
            prev2 = tmp;
        }

        return std::max(prev1, prev2);
    }
};